File: matlab/exercise_01.m
% BCI workshop 2015
% Exercise 1: A neurofeedback interface (single-channel)
% 
% Description:
% In this exercise, we'll try and play around with a simple interface that
% receives EEG from one electrode, computes standard frequency band powers 
% and displays both the raw signals and the features.
% 

clear;
close all;

addpath(genpath('bci_workshop_tools\'));
global stop_loop
stop_loop = false;

% MuLES connection parameters    
mules_ip = '127.0.0.1';
muse_port = 30000;

% Creates a mules_client
mules_client = MulesClient(mules_ip, muse_port);
params = mules_client.getparams();

%% Set the experiment parameters
    
eeg_buffer_secs = 15;  % Size of the EEG data buffer used for plotting the 
                       % signal (in seconds) 
win_test_secs = 1;     % Length of the window used for computing the features 
                       % (in seconds)
overlap_secs = 0.5;    % Overlap between two consecutive windows (in seconds)
shift_secs = win_test_secs - overlap_secs;
index_channel = 1;     % Index of the channnel to be used (with the Muse, we 
                       % can choose from 0 to 3)

% This line changes params to work with only one electrode
params{6} = {'CH1','STATUS'}';
name_of_channels   = params{6};
sampling_frequency = params{3};

% Get name of features
names_of_features = feature_names(name_of_channels);

%% Initialize the buffers for storing raw EEG and features

% Initialize raw EEG data buffer (for plotting)
eeg_buffer = zeros(sampling_frequency * eeg_buffer_secs , numel(name_of_channels)); 

% Compute the number of windows in "eeg_buffer_secs" (used for plotting)
n_win_test = floor((eeg_buffer_secs - win_test_secs) / shift_secs) + 1;

% Initialize the feature data buffer (for plotting)
feat_buffer = zeros(n_win_test , numel(names_of_features));

% Initialize the plots
h_eeg_fig = figure('CloseRequestFcn',@fig_closereq);
h_eeg_ax = axes();
figure();
h_feat_ax = axes();


%% Start pulling data
mules_client.flushdata();  % Flush old data from MuLES       
tone(500,500); % Beep sound
             
disp(' Close the raw EEG signal figure window to break the While Loop');

while ~stop_loop
    % 1- ACQUIRE DATA 
    eeg_data = mules_client.getdata(shift_secs, false); % Obtain EEG data from MuLES  
    eeg_data = eeg_data(:,[index_channel, end]); % Keep only one electrode (and the STATUS channel) for further analysis      
    eeg_buffer = updatebuffer(eeg_buffer, eeg_data); % Update EEG buffer

    % 2- COMPUTE FEATURES 
    % Get newest samples from the buffer 
    data_window = getlastdata(eeg_buffer, win_test_secs * sampling_frequency);
    % Compute features on "data_window" 
    feat_vector = compute_feature_vector(data_window, sampling_frequency);
    feat_buffer = updatebuffer(feat_buffer, feat_vector); % Update the feature buffer

    % 3- VISUALIZE THE RAW EEG AND THE FEATURES        
    plot_channels(h_eeg_ax, eeg_buffer, sampling_frequency, name_of_channels);
    plot_channels(h_feat_ax, feat_buffer, 1/shift_secs, names_of_features);
    
    pause(0.00001);   
     
end

% The raw EEG signal figure is repltoted
figure()
h_eeg_ax = axes();
plot_channels(h_eeg_ax, eeg_buffer, sampling_frequency, name_of_channels);

% Close connection with MuLES
mules_client.disconnect(); % Close connection

rmpath(genpath('bci_workshop_tools\'));

File: python/exercise_01.py
# -*- coding: utf-8 -*-
"""
Exercise 1: A neurofeedback interface (single-channel)
======================================================

Description:
In this exercise, we'll try and play around with a simple interface that
receives EEG from one electrode, computes standard frequency band powers
and displays both the raw signals and the features.

"""

import numpy as np  # Module that simplifies computations on matrices
import matplotlib.pyplot as plt  # Module used for plotting
from pylsl import StreamInlet, resolve_byprop  # Module to receive EEG data

import bci_workshop_tools as BCIw  # Our own functions for the workshop


if __name__ == "__main__":

    """ 1. CONNECT TO EEG STREAM """

    # Search for active LSL stream
    print('Looking for an EEG stream...')
    streams = resolve_byprop('type', 'EEG', timeout=2)
    if len(streams) == 0:
        raise RuntimeError('Can\'t find EEG stream.')

    # Set active EEG stream to inlet and apply time correction
    print("Start acquiring data")
    inlet = StreamInlet(streams[0], max_chunklen=12)
    eeg_time_correction = inlet.time_correction()

    # Get the stream info and description
    info = inlet.info()
    description = info.desc()

    # Get the sampling frequency
    # This is an important value that represents how many EEG data points are
    # collected in a second. This influences our frequency band calculation.
    fs = int(info.nominal_srate())

    """ 2. SET EXPERIMENTAL PARAMETERS """

    # Length of the EEG data buffer (in seconds)
    # This buffer will hold last n seconds of data and be used for calculations
    buffer_length = 15

    # Length of the epochs used to compute the FFT (in seconds)
    epoch_length = 1

    # Amount of overlap between two consecutive epochs (in seconds)
    overlap_length = 0.8

    # Amount to 'shift' the start of each next consecutive epoch
    shift_length = epoch_length - overlap_length

    # Index of the channel (electrode) to be used
    # 0 = left ear, 1 = left forehead, 2 = right forehead, 3 = right ear
    index_channel = [0]
    ch_names = ['ch1']  # Name of our channel for plotting purposes

    # Get names of features
    # ex. ['delta - CH1', 'pwr-theta - CH1', 'pwr-alpha - CH1',...]
    feature_names = BCIw.get_feature_names(ch_names)

    """ 3. INITIALIZE BUFFERS """

    # Initialize raw EEG data buffer (for plotting)
    eeg_buffer = np.zeros((int(fs * buffer_length), 1))
    filter_state = None  # for use with the notch filter

    # Compute the number of epochs in "buffer_length" (used for plotting)
    n_win_test = int(np.floor((buffer_length - epoch_length) /
                              shift_length + 1))

    # Initialize the feature data buffer (for plotting)
    feat_buffer = np.zeros((n_win_test, len(feature_names)))

    # Initialize the plots
    plotter_eeg = BCIw.DataPlotter(fs * buffer_length, ch_names, fs)
    plotter_feat = BCIw.DataPlotter(n_win_test, feature_names,
                                    1 / shift_length)

    """ 3. GET DATA """

    # The try/except structure allows to quit the while loop by aborting the
    # script with <Ctrl-C>
    print('Press Ctrl-C in the console to break the while loop.')

    try:
        # The following loop does what we see in the diagram of Exercise 1:
        # acquire data, compute features, visualize raw EEG and the features
        while True:

            """ 3.1 ACQUIRE DATA """
            # Obtain EEG data from the LSL stream
            eeg_data, timestamp = inlet.pull_chunk(
                    timeout=1, max_samples=int(shift_length * fs))

            # Only keep the channel we're interested in
            ch_data = np.array(eeg_data)[:, index_channel]

            # Update EEG buffer
            eeg_buffer, filter_state = BCIw.update_buffer(
                    eeg_buffer, ch_data, notch=True,
                    filter_state=filter_state)

            """ 3.2 COMPUTE FEATURES """
            # Get newest samples from the buffer
            data_epoch = BCIw.get_last_data(eeg_buffer,
                                            epoch_length * fs)

            # Compute features
            feat_vector = BCIw.compute_feature_vector(data_epoch, fs)
            feat_buffer, _ = BCIw.update_buffer(feat_buffer,
                                                np.asarray([feat_vector]))

            """ 3.3 VISUALIZE THE RAW EEG AND THE FEATURES """
            plotter_eeg.update_plot(eeg_buffer)
            plotter_feat.update_plot(feat_buffer)
            plt.pause(0.00001)

    except KeyboardInterrupt:
        print('Closing!')
